File: CSharp/SpaceVIL/SpaceVIL/Engine/WindowProcessor.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Runtime.InteropServices;
using Glfw3;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal class WindowProcessor
    {
        private Geometry _geometryForRestore = new Geometry();
        private Position _positionForRestore = new Position();
        private Glfw.Monitor _monitor;
        private Glfw.VideoMode _vid;
        CommonProcessor _commonProcessor;
        internal WindowProcessor(CommonProcessor commonProcessor)
        {
            _commonProcessor = commonProcessor;
        }

        internal void SetWindowSize(int width, int height, Scale scale)
        {
            if (WindowsBox.GetCurrentFocusedWindow() != _commonProcessor.Window)
                Glfw.FocusWindow(_commonProcessor.Window.GetGLWID());

            if (_commonProcessor.Window.IsKeepAspectRatio)
            {
                float currentW = width;
                float currentH = height;
                float ratioW = _commonProcessor.Window.RatioW;
                float ratioH = _commonProcessor.Window.RatioH;
                float xScale = (currentW / ratioW);
                float yScale = (currentH / ratioH);
                float ratio = 0;
                Side handlerContainerSides = _commonProcessor.RootContainer._sides;
                if (handlerContainerSides.HasFlag(Side.Left)
                        || handlerContainerSides.HasFlag(Side.Right))
                    ratio = xScale;
                else
                    ratio = yScale;

                width = (int)(ratioW * ratio);
                height = (int)(ratioH * ratio);
            }

            if (width > _commonProcessor.Window.GetMaxWidth())
                width = _commonProcessor.Window.GetMaxWidth();

            if (height > _commonProcessor.Window.GetMaxHeight())
                height = _commonProcessor.Window.GetMaxHeight();

            Glfw.SetWindowSize(_commonProcessor.Handler.GetWindowId(),
                    (int)(width * scale.GetXScale()), (int)(height * scale.GetYScale()));

            _commonProcessor.Events.SetEvent(InputEventType.WindowResize);
        }

        internal void SetWindowPos(int x, int y)
        {
            Glfw.SetWindowPos(_commonProcessor.Handler.GetWindowId(), x, y);
            _commonProcessor.Events.SetEvent(InputEventType.WindowMove);
        }

        internal void Drop(Int64 glfwwnd, int count, string[] paths)
        {
            DropArgs dargs = new DropArgs();
            dargs.Count = count;
            dargs.Paths = new List<String>(paths);
            dargs.Item = _commonProcessor.HoveredItem;
            _commonProcessor.Manager.AssignActionsForSender(InputEventType.WindowDrop, dargs, _commonProcessor.RootContainer,
                _commonProcessor.UnderFocusedItems, false);
        }

        internal void MinimizeWindow()
        {
            _commonProcessor.InputLocker = true;
            _commonProcessor.Events.SetEvent(InputEventType.WindowMinimize);
            Glfw.IconifyWindow(_commonProcessor.Handler.GetWindowId());
            _commonProcessor.InputLocker = false;
        }

        internal void MaximizeWindow(Scale scale)
        {
            _commonProcessor.InputLocker = true;
            if (_commonProcessor.Window.IsMaximized)
            {
                Glfw.RestoreWindow(_commonProcessor.Handler.GetWindowId());
                _commonProcessor.Events.SetEvent(InputEventType.WindowRestore);
                _commonProcessor.Window.IsMaximized = false;
            }
            else
            {
                Glfw.MaximizeWindow(_commonProcessor.Handler.GetWindowId());
                _commonProcessor.Events.SetEvent(InputEventType.WindowMaximize);
                _commonProcessor.Window.IsMaximized = true;
            }

            int width = _commonProcessor.Window.GetWidth();
            int height = _commonProcessor.Window.GetHeight();

            Glfw.SetWindowSize(_commonProcessor.Handler.GetWindowId(),
                    (int)(width * scale.GetXScale()), (int)(height * scale.GetYScale()));

            _commonProcessor.InputLocker = false;
        }

        internal void FullScreenWindow()
        {
            _commonProcessor.InputLocker = true;
            _monitor = Glfw.GetPrimaryMonitor();
            if (!_commonProcessor.Window.IsFullScreen)
            {
                _vid = Glfw.GetVideoMode(_monitor);
                _geometryForRestore.SetSize(_commonProcessor.Window.GetWidth(), _commonProcessor.Window.GetHeight());
                _positionForRestore.SetPosition(_commonProcessor.Window.GetX(), _commonProcessor.Window.GetY());
                Glfw.SetWindowMonitor(_commonProcessor.Handler.GetWindowId(), _monitor, 0, 0, _vid.Width, _vid.Height, _vid.RefreshRate);
                Glfw.FocusWindow(_commonProcessor.Handler.GetWindowId());
                _commonProcessor.Window.SetWidthDirect(_vid.Width);
                _commonProcessor.Window.SetHeightDirect(_vid.Height);
                _commonProcessor.Window.IsFullScreen = true;
            }
            else
            {
                Glfw.SetWindowMonitor(_commonProcessor.Handler.GetWindowId(), Glfw.Monitor.None,
                        _positionForRestore.GetX(), _positionForRestore.GetY(),
                        _geometryForRestore.GetWidth(), _geometryForRestore.GetHeight(), 0);
                _commonProcessor.Window.SetWidthDirect(_geometryForRestore.GetWidth());
                _commonProcessor.Window.SetHeightDirect(_geometryForRestore.GetHeight());
                _commonProcessor.Window.IsFullScreen = false;
            }
            _commonProcessor.InputLocker = false;
        }

        internal void Focus(Int64 wnd, bool value)
        {
            if (!_commonProcessor.Handler.Focusable)
                return;

            _commonProcessor.Events.ResetAllEvents();
            _commonProcessor.Tooltip.InitTimer(false);

            if (value)
            {
                if (_commonProcessor.Handler.Focusable)
                {
                    WindowsBox.SetCurrentFocusedWindow(_commonProcessor.Window);
                    _commonProcessor.Window.SetFocus(value);
                    _commonProcessor.Handler.Focused = value;
                }
            }
            else
            {
                if (_commonProcessor.Window.IsDialog)
                {
                    _commonProcessor.Window.SetFocus(true);
                    _commonProcessor.Handler.Focused = true;
                }
                else
                {
                    _commonProcessor.Window.SetFocus(value);
                    _commonProcessor.Handler.Focused = value;
                    
                    if (_commonProcessor.Window.IsOutsideClickClosable)
                    {
                        _commonProcessor.ResetItems();
                        _commonProcessor.Window.Close();
                    }
                }
            }
        }

        private Glfw.Image _iconBig;
        private Glfw.Image _iconSmall;

        internal void SetIcons(Bitmap ibig, Bitmap ismall)
        {
            if (_iconBig.Pixels == null)
            {
                _iconBig.Width = ibig.Width;
                _iconBig.Height = ibig.Height;
                _iconBig.Pixels = GetImagePixels(ibig);
            }
            if (_iconSmall.Pixels == null)
            {
                _iconSmall.Width = ismall.Width;
                _iconSmall.Height = ismall.Height;
                _iconSmall.Pixels = GetImagePixels(ismall);
            }
        }

        internal void ApplyIcon()
        {
            if ((_iconBig.Pixels == null) || (_iconSmall.Pixels == null))
                return;
            Glfw.Image[] images = new Glfw.Image[2];
            images[0] = _iconBig;
            images[1] = _iconSmall;
            Glfw.SetWindowIcon(_commonProcessor.Handler.GetWindowId(), images);
        }

        private byte[] GetImagePixels(Bitmap icon)
        {
            if (icon == null)
                return null;

            // List<byte> _map = new List<byte>();
            // Bitmap bmp = new Bitmap(icon);
            // for (int j = 0; j < icon.Height; j++)
            // {
            //     for (int i = 0; i < icon.Width; i++)
            //     {
            //         Color pixel = bmp.GetPixel(i, j);
            //         _map.Add(pixel.R);
            //         _map.Add(pixel.G);
            //         _map.Add(pixel.B);
            //         _map.Add(pixel.A);
            //     }
            // }
            // return _map.ToArray();

            int imageWidth = icon.Width;
            int imageHeight = icon.Height;
            int size = imageWidth * imageHeight;
            BitmapData bitData = icon.LockBits(
                new System.Drawing.Rectangle(System.Drawing.Point.Empty, icon.Size), ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);
            int[] pixels = new int[size];
            Marshal.Copy(bitData.Scan0, pixels, 0, size);
            icon.UnlockBits(bitData);
            byte[] result = new byte[size * 4];
            int index = 0;
            foreach (int pixel in pixels)
            {
                byte a = (byte)((pixel & 0xFF000000) >> 24);
                byte r = (byte)((pixel & 0x00FF0000) >> 16);
                byte g = (byte)((pixel & 0x0000FF00) >> 8);
                byte b = (byte)(pixel & 0x000000FF);
                result[index++] = r;
                result[index++] = g;
                result[index++] = b;
                result[index++] = a;
            }
            return result;
        }

        internal static void ToggleToFullScreenMode(CoreWindow window)
        {
            Glfw.Monitor monitor = Glfw.GetPrimaryMonitor();
            Glfw.VideoMode vid = Glfw.GetVideoMode(monitor);
            Glfw.SetWindowMonitor(window.GetGLWID(), monitor, 0, 0, vid.Width, vid.Height, vid.RefreshRate);
        }
        internal static void ToggleToWindowedMode(Int64 window)
        {
            Glfw.SetWindowMonitor(window, Glfw.Monitor.None, 0, 0, 500, 300, 60);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/WindowProcessor.java
package com.spvessel.spacevil;

import java.util.LinkedList;
import java.util.List;
import java.awt.image.BufferedImage;

import com.spvessel.spacevil.Common.CommonService;
import com.spvessel.spacevil.Core.DropArgs;
import com.spvessel.spacevil.Core.Geometry;
import com.spvessel.spacevil.Core.Position;
import com.spvessel.spacevil.Core.Scale;
import com.spvessel.spacevil.Flags.InputEventType;
import com.spvessel.spacevil.Flags.OSType;
import com.spvessel.spacevil.Flags.Side;

import com.spvessel.spacevil.internal.Wrapper.*;

final class WindowProcessor {
    private GlfwWrapper glfw = null;

    private Geometry _geometryForRestore = new Geometry();
    private Position _positionForRestore = new Position();
    private long _monitor;
    private GLFWVidMode _vid;
    private CommonProcessor _commonProcessor;

    WindowProcessor(CommonProcessor processor) {
        glfw = GlfwWrapper.get();
        _commonProcessor = processor;
    }

    void setWindowSize(int width, int height, Scale scale) {
        if (WindowsBox.getCurrentFocusedWindow() != _commonProcessor.window)
            glfw.FocusWindow(_commonProcessor.window.getGLWID());

        if (_commonProcessor.window.isKeepAspectRatio) {
            float currentW = width;
            float currentH = height;

            float ratioW = _commonProcessor.window.ratioW;
            float ratioH = _commonProcessor.window.ratioH;

            float xScale = (currentW / ratioW);
            float yScale = (currentH / ratioH);

            float ratio = 0;

            List<Side> handlerContainerSides = _commonProcessor.rootContainer.getSides();

            if (handlerContainerSides.contains(Side.Left) || handlerContainerSides.contains(Side.Right))
                ratio = xScale;
            else
                ratio = yScale;

            width = (int) (ratioW * ratio);
            height = (int) (ratioH * ratio);
        }

        if (CommonService.getOSType() != OSType.Mac) {
            if (width > _commonProcessor.window.getMaxWidth())
                width = _commonProcessor.window.getMaxWidth();

            if (height > _commonProcessor.window.getMaxHeight())
                height = _commonProcessor.window.getMaxHeight();

            glfw.SetWindowSize(_commonProcessor.handler.getWindowId(), (int) (width * scale.getXScale()),
                    (int) (height * scale.getYScale()));
        } else {
            glfw.SetWindowSize(_commonProcessor.handler.getWindowId(), width, height);
        }

        _commonProcessor.events.setEvent(InputEventType.WindowResize);
    }

    void setWindowPos(int x, int y) {
        glfw.SetWindowPos(_commonProcessor.handler.getWindowId(), x, y);
        _commonProcessor.events.setEvent(InputEventType.WindowMove);
    }

    void drop(long window, int count, String[] paths) {
        DropArgs dargs = new DropArgs();
        dargs.count = count;
        dargs.paths = new LinkedList<>();
        dargs.item = _commonProcessor.hoveredItem;

        for (int i = 0; i < count; i++) {
            dargs.paths.add(paths[i]);
        }
        _commonProcessor.manager.assignActionsForSender(InputEventType.WindowDrop, dargs,
                _commonProcessor.rootContainer, _commonProcessor.underFocusedItems, false);
    }

    void minimizeWindow() {
        _commonProcessor.inputLocker = true;
        _commonProcessor.events.setEvent(InputEventType.WindowMinimize);
        if (!_commonProcessor.window.isMinimized()) {
            glfw.IconifyWindow(_commonProcessor.handler.getWindowId());
            _commonProcessor.window.setMinimized(true);
        } else {
            glfw.RestoreWindow(_commonProcessor.handler.getWindowId());
            _commonProcessor.window.setMinimized(false);
        }
        _commonProcessor.inputLocker = false;
    }

    void maximizeWindow(Scale scale) {
        _commonProcessor.inputLocker = true;

        if (_commonProcessor.window.isMaximized) {
            glfw.RestoreWindow(_commonProcessor.handler.getWindowId());
            _commonProcessor.events.setEvent(InputEventType.WindowRestore);
            _commonProcessor.window.isMaximized = false;
        } else {
            glfw.MaximizeWindow(_commonProcessor.handler.getWindowId());
            _commonProcessor.events.setEvent(InputEventType.WindowMaximize);
            _commonProcessor.window.isMaximized = true;
        }

        if (CommonService.getOSType() != OSType.Mac) {
            int width = _commonProcessor.window.getWidth();
            int height = _commonProcessor.window.getHeight();
            glfw.SetWindowSize(_commonProcessor.handler.getWindowId(), (int) (width * scale.getXScale()),
                    (int) (height * scale.getYScale()));
        }

        _commonProcessor.inputLocker = false;
    }

    void fullScreenWindow() {
        _commonProcessor.inputLocker = true;
        _monitor = glfw.GetPrimaryMonitor();
        if (!_commonProcessor.window.isFullScreen) {
            _vid = glfw.GetVideoMode(_monitor);
            _geometryForRestore.setSize(_commonProcessor.window.getWidth(), _commonProcessor.window.getHeight());
            _positionForRestore.setPosition(_commonProcessor.window.getX(), _commonProcessor.window.getY());
            glfw.SetWindowMonitor(_commonProcessor.handler.getWindowId(), _monitor, 0, 0, _vid.width, _vid.height,
                    _vid.refreshRate);
            _commonProcessor.window.setWidthDirect(_vid.width);
            _commonProcessor.window.setHeightDirect(_vid.height);
            _commonProcessor.window.isFullScreen = true;
        } else {
            glfw.SetWindowMonitor(_commonProcessor.handler.getWindowId(), 0, _positionForRestore.getX(),
                    _positionForRestore.getY(), _geometryForRestore.getWidth(), _geometryForRestore.getHeight(), 0);
            _commonProcessor.window.setWidthDirect(_geometryForRestore.getWidth());
            _commonProcessor.window.setHeightDirect(_geometryForRestore.getHeight());
            _commonProcessor.window.isFullScreen = false;
        }
        _commonProcessor.inputLocker = false;
    }

    void focus(long wnd, boolean value) {
        if (!_commonProcessor.handler.focusable)
            return;

        _commonProcessor.events.resetAllEvents();
        _commonProcessor.toolTip.initTimer(false);

        if (value) {
            if (_commonProcessor.handler.focusable) {
                WindowsBox.setCurrentFocusedWindow(_commonProcessor.window);
                _commonProcessor.window.setFocus(value);
                _commonProcessor.handler.focused = value;
            }
        } else {
            if (_commonProcessor.window.isDialog) {
                _commonProcessor.window.setFocus(true);
                _commonProcessor.handler.focused = true;
            } else {
                _commonProcessor.window.setFocus(value);
                _commonProcessor.handler.focused = value;

                if (_commonProcessor.window.isOutsideClickClosable) {
                    _commonProcessor.resetItems();
                    _commonProcessor.window.close();
                }
            }
        }
    }

    private byte[] createByteImage(BufferedImage image) {
        // List<Byte> bmp = new LinkedList<Byte>();
        // for (int j = 0; j < image.getHeight(); j++) {
        // for (int i = 0; i < image.getWidth(); i++) {
        // byte[] bytes = ByteBuffer.allocate(4).putInt(image.getRGB(i, j)).array();
        // bmp.add(bytes[1]);
        // bmp.add(bytes[2]);
        // bmp.add(bytes[3]);
        // bmp.add(bytes[0]);
        // }
        // }
        // ByteBuffer result = BufferUtils.createByteBuffer(bmp.size());
        // for (byte var : bmp) {
        // result.put(var);
        // }
        // result.rewind();
        return VramTexture.getByteBuffer(image);
    }

    void applyIcon(BufferedImage iconBig, BufferedImage iconSmall) {
        if ((iconBig == null) || (iconSmall == null))
            return;
        GLFWImage[] gb = new GLFWImage[2];

        GLFWImage s = new GLFWImage();
        s.width = iconSmall.getWidth();
        s.height = iconSmall.getHeight();
        s.pixels = createByteImage(iconSmall);

        GLFWImage b = new GLFWImage();
        b.width = iconBig.getWidth();
        b.height = iconBig.getHeight();
        b.pixels = createByteImage(iconBig);

        gb[0] = s;
        gb[1] = b;

        glfw.SetWindowIcon(_commonProcessor.handler.getWindowId(), 2, gb);
    }
}